include/picongpu/plugins/transitionRadiation/Calculator.hpp

```diff
@@ -0,0 +1,238 @@
+            float3_X const lookDirection
+        ) : 
+            particle( particleSet ),
+            lookDirection( lookDirection ),
+            parMomPhi( particle.getMomPhi( ) ),
+            detectorPhi( 
+                picongpu::math::atan2(
+                    lookDirection.z( ), 
+                    lookDirection.x( )
+                ) + picongpu::PI
+            ),
+            parSqrtOnePlusUSquared( 
+                picongpu::math::sqrt( 1 + particle.getU( ) * particle.getU( ) ) 
+            )
+        { 
+            // Frequent calculations
+            // Momentum Space for Particle:
+            picongpu::math::sincos( 
+                particle.getMomTheta( ), 
+                parMomSinTheta, 
+                parMomCosTheta 
+            );
+            picongpu::math::sincos( 
+                parMomPhi - detectorPhi, 
+                parMomSinPhi, 
+                parMomCosPhi
+            );
+            
+            // Detector Position
+            float_X const detectorTheta = picongpu::math::acos( 
+                lookDirection.y( ) / picongpu::math::sqrt( ( 
+                    lookDirection * lookDirection
+                ).sumOfComponents( ) )
+            );
+            picongpu::math::sincos( 
+                detectorTheta, 
+                detectorSinTheta, 
+                detectorCosTheta 
+            );
+        }
+
+        /** Perpendicular part of normalized energy
+         * 
+         * Calculates perpendicular part to movement direction of normalized energy
+         * determined by formula:
+         * @f[E_{perp} = (u^2 \cos{\psi} \sin{\psi} \sin{\phi} \cos{\theta}) / 
+         *          ((\sqrt{1 + u^2} - u \sin{\psi} \cos{\phi} \sin{\theta})^2 - u^2 \cos{\phi}^2 \cos{\theta}^2)@f]
+         * where Psi is the azimuth angle of the particle momentum and theta is
+         * the azimuth angle of the detector position to the movement direction y
+         * 
+         * @return perpendicular part of normalized energy
+         */
+        HDINLINE 
+        float_X 
+        calcEnergyPerp( ) const
+        {
+            float_X const uSquared = particle.getU( ) * particle.getU( );
+            float_X const a = uSquared * parMomCosTheta * parMomSinTheta * 
+                parMomSinPhi * detectorCosTheta;
+
+            // Denominator
+            float_X const x = parSqrtOnePlusUSquared - 
+                particle.getU( ) * parMomSinTheta * parMomCosPhi * detectorSinTheta;
+            float_X const y = particle.getU( ) * parMomCosTheta * detectorCosTheta;
+
+            float_X denominator = x * x - y * y;
+
+            // Preventing division by 0
+            if( math::abs( denominator ) < DIV_BY_ZERO_MINIMUM )
+            {
+                if( denominator < 0.0 )
+                    denominator = -DIV_BY_ZERO_MINIMUM;
+                else
+                    denominator = DIV_BY_ZERO_MINIMUM;
+            }
```

Comment:
These `if` clauses might light to branching of the parallel code and thus to a slowdown of the parallel execution.

Please consider rewriting this as:
```c++
const bool prevent_div_by_zero = math::abs( denominator ) < DIV_BY_ZERO_MINIMUM;
    denominator = int(!prevent_div_by_zero) * denominator 
        + int(prevent_div_by_zero) * ( -2 * int( denominator < 0.0 ) + 1) * DIV_BY_ZERO_MINIMUM;
```

Reply:
@psychocoderHPC Could you please comment whether such optimizations are useful here?

Reply:
I have included
```cpp
const bool preventDivByZero = math::abs( denominator ) < DIV_BY_ZERO_MINIMUM;
denominator = int( !preventDivByZero ) * denominator +  int( preventDivByZero ) * ( -2 * int( denominator < 0.0 ) + 1 ) * DIV_BY_ZERO_MINIMUM;
```
in `calcEnergyPara` and `calcEnergyPerp` and 
```cpp
const bool longMovingParticle = exponent.get_real() == -1.0;
return float_X( longMovingParticle ) * complex_X( 0.0, 0.0 ) +  float_X( !longMovingParticle ) * complex_X( math::exp( exponent * omega ) );
```
in the [formfactor](https://github.com/ComputationalRadiationPhysics/picongpu/pull/3003#discussion_r307182735) for the `Transition Radiation Example` setup with a bit of a higher resolution for a quick benchmark. The `calculation simulation time` for these 5 consecutive transition radiation calculations with my original code was `18 minutes` long, where it was  `19 minutes long` with all the three changes. I will quickly run the simulation again with my code in the formfactor and your code in the energies to see how this effects the simulation time.

Reply:
Now I have 2 more simulations with 1) the formfactor-change and 2) the change in the energies. The simulation with the formfactor change from @PrometheusPi takes `17:44 minutes` and the second simulation takes `19:40 minutes`. These benchmarks are just rough approximations, but I think they should be a good estimate for the amount of time which can be saved with these changes. I run one of the simulations twice and the `calclulation simulation time` was identical. Hence I would use [this](https://github.com/ComputationalRadiationPhysics/picongpu/pull/3003#discussion_r307182735) method, but not the one from the energy functions.